libevmasm/RuleList.h

```diff
@@ -374,6 +390,30 @@ std::vector<SimplificationRule<Pattern>> simplificationRuleListPart7(
 		false
 	});
 
+	rules.push_back({
+		// AND(A, SHR(B, X)) -> SHR(B, X) if 2^B-1 == A
+		{Instruction::AND, {A, {Instruction::SHR, {B, X}}}},
+		[=]() -> Pattern { return {Instruction::SHR, {B, X}}; },
+		false,
+		[=]() {
+			if (B.d() > 256)
+				return false;
+			return (u256(2) << (256 - static_cast<int>(B.d())-1)) - 1 == A.d();
```

Comment:
What about `A.d() & (u256(-1) >> B.d()) == A.d()`?

Reply:
I would at least change this to `u256(1) <<` instead of starting with 2 and then shifting one less...

Reply:
We will probably get a problem if `B.d()` is zero.

Reply:
The rule passes the tests, too. Changing..

Reply:
As said in the comment above, I think it needs to be ``A.d() & (u256(-1) >> B.d()) ==  (u256(-1) >> B.d())``.

Your suggestion would be true for A=1 and B < 255, e.g. ``AND(1, SHR(2,X)) -> SHR(2,X)`` which would be wrong.

Reply:
Maybe there's not enough tests for this then :-)?

Reply:
Ah right, it's `== -1 >> B`!

Reply:
What passes was `(u256(-1) >> static_cast<int>(A.d())) == B.d();`, didn't try the one with `A.d() &`

I think for the case `B.d() == 0` another optimizer rule will remove the whole SHL anyway. But it won't hurt to explicitly exclude that case

Reply:
As mentioned in the comment above ``B.d() == 0`` is just fine, if we have a condition that's correct in general :-).